xl-v1.go

```diff
@@ -375,47 +374,75 @@ func (xl XL) ListFiles(volume, prefix, marker string, recursive bool, count int)
-	if err != nil {
-		return nil, true, err
-	}
-
-	for _, fsFileInfo := range fsFilesInfo {
-		// Skip metadata files.
-		if strings.HasSuffix(fsFileInfo.Name, metadataFile) {
-			continue
+	for {
+		fsFilesInfo, eof, err = disk.ListFiles(volume, prefix, markerPath, recursive, count)
+		if err != nil {
+			return nil, true, err
 		}
-		var fileInfo FileInfo
-		var isLeaf bool
-		if fsFileInfo.Mode.IsDir() {
-			isLeaf = xl.isLeafDirectory(volume, fsFileInfo.Name)
+		for _, fsFileInfo := range fsFilesInfo {
+			// Skip metadata files.
+			if strings.HasSuffix(fsFileInfo.Name, metadataFile) {
+				continue
+			}
+			var fileInfo FileInfo
+			var isLeaf bool
+			if fsFileInfo.Mode.IsDir() {
+				isLeaf = xl.isLeafDirectory(volume, fsFileInfo.Name)
+			}
+			if isLeaf || !fsFileInfo.Mode.IsDir() {
+				// Extract the parent of leaf directory or file to get the
+				// actual name.
+				path := slashpath.Dir(fsFileInfo.Name)
+				fileInfo, err = xl.extractFileInfo(volume, path)
+				if err != nil {
+					// For a leaf directory, if err is FileNotFound then
+					// perhaps has a missing metadata. Ignore it and let
+					// healing finish its job it will become available soon.
+					if err == errFileNotFound {
+						continue
+					}
+					// For any other errors return to the caller.
+					return nil, true, err
+				}
+			} else {
+				fileInfo = fsFileInfo
+			}
+			filesInfo = append(filesInfo, fileInfo)
+			count--
+			if count == 0 {
+				break
+			}
+		}
+		lenFsFilesInfo := len(fsFilesInfo)
+		if lenFsFilesInfo != 0 {
+			markerPath = fsFilesInfo[lenFsFilesInfo-1].Name
 		}
-		if isLeaf || !fsFileInfo.Mode.IsDir() {
-			// Extract the parent of leaf directory or file to get the
-			// actual name.
-			path := slashpath.Dir(fsFileInfo.Name)
-			fileInfo, err = xl.extractFileInfo(volume, path)
+		if count == 0 && recursive && !strings.HasSuffix(markerPath, metadataFile) {
+			// If last entry is not part.json then loop once more to check if we
+			// have reached eof.
+			fsFilesInfo, eof, err = disk.ListFiles(volume, prefix, markerPath, recursive, 1)
 			if err != nil {
-				// For a leaf directory, if err is FileNotFound then
-				// perhaps has a missing metadata. Ignore it and let
-				// healing finish its job it will become available soon.
-				if err == errFileNotFound {
-					continue
-				}
-				// For any other errors return to the caller.
 				return nil, true, err
 			}
-		} else {
-			fileInfo = fsFileInfo
+			if !eof {
+				// part.N and part.json are always in pairs and hence this
+				// entry has to be part.json. If not better to manually investigate
+				// and fix it.
+				// For the next ListFiles() call we can safely assume that the
+				// marker is "object/part.json"
+				if !strings.HasSuffix(fsFilesInfo[0].Name, metadataFile) {
+					return nil, true, errUnexpected
+				}
+			}
+		}
+		if count == 0 || eof {
+			break
 		}
-		filesInfo = append(filesInfo, fileInfo)
 	}
-	sort.Sort(byFileInfoName(filesInfo))
```

Comment:
But such checks should be part of our test suite. I see that there is minimal test case for it:
```
        // check ordering of pages                                                                                                                                                
        {                                                                                                                                                                         
                result, err = obj.ListObjects("bucket", "", "", "", 1000)                                                                                                         
                c.Assert(err, check.IsNil)                                                                                                                                        
                c.Assert(result.Objects[0].Name, check.Equals, "newPrefix")                                                                                                       
                c.Assert(result.Objects[1].Name, check.Equals, "newPrefix2")                                                                                                      
                c.Assert(result.Objects[2].Name, check.Equals, "obj0")                                                                                                            
                c.Assert(result.Objects[3].Name, check.Equals, "obj1")                                                                                                            
                c.Assert(result.Objects[4].Name, check.Equals, "obj10")                                                                                                           
        }
```

but a more exhaustive test case to check if sorted list is returned should be done.

If verification is really needed then better do it at the object-layer so that we can be sure that s3-client always gets a sorted list. But still this verification is better done by the test suite. Let me know.